splitio/javascript-client
# [WIP] Feature/imp dedupe

# JS SDK

## What did you accomplish?
 - Added impressions dedupe logic to avoid sending duplicated impressions:
    - Added `ImpressionsCount` cache to count impressions generated per feature.
    - Added `ImpressionObserver` to detect previous generated impression.
    - Added new endpoint (`testImpressions/count`) to send all the generated impressions in a time frame.
    - Added beacon endpoint for ImpressionsCount.
    - Added murmur3 128 bits hashing and splitted logic between browser and node.
    - Added `OPTIMIZED` and `DEBUG` modes in order to enabling/disabling how impressions are going to be sent into Split servers,
        - `OPTIMIZED`: will send unique impressions in a timeframe in order to reduce how many times impressions are posted to Split.
        - `DEBUG`: will send every impression generated to Split.
 - Updated JSON bulk properties for Impressions to decrease the payload size.
 - Merged `ImpressionsTracker` and `ImpressionTracker` to be used in both type of evaluations: `getTreatment` and `getTreatments`.
 - Fixed vulnerability for `node-fetch` package. 

## How do we test the changes introduced in this PR?
- Run `npm test`
- SDK will run in Optimized mode, you will see less post made to the BE server. If you run in browser mode, under the tab Network, you will see new endpoint that sends all the generated impressions count.

## Extra Notes
